src/controllers/api-key.js

```diff
@@ -10,82 +10,259 @@
+      }
+    });
+
+    if (!hasText(data.name)) {
+      throw new ErrorWithStatusCode('Invalid request: missing name in the request data', STATUS_BAD_REQUEST);
+    }
+  }
+
+  /**
+   * Validate the IMS Org ID.
+   * @param {string} imsOrgId - The IMS Organization ID of the user.
+   * @param {string} imsUserToken - The IMS User access token provided by the user.
+   * @throws {ErrorWithStatusCode} If the IMS Org ID is invalid or
+   * if the user does not belong to the given imsOrg.
+   * @returns {object} imsUserProfile - The IMS User profile.
+   */
+  async function validateImsOrgId(imsOrgId, imsUserToken) {
+    if (!hasText(imsOrgId)) {
+      throw new ErrorWithStatusCode('Missing x-gw-ims-org-id header', STATUS_UNAUTHORIZED);
+    }
+    const imsUserProfile = await imsClient.getImsUserProfile(imsUserToken);
+    const { organizations } = imsUserProfile;
+    if (!organizations.includes(imsOrgId)) {
+      throw new ErrorWithStatusCode('Invalid request: Unable to find a reference to the Organization provided.', STATUS_UNAUTHORIZED);
+    }
+    return imsUserProfile;
   }
 
-  // eslint-disable-next-line no-unused-vars
-  function validateImsOrgId(imsOrgId) {
-    // Not implemented yet
+  /**
+   * Get the IMS user token from the headers.
+   * @param {object} headers - The headers of the request.
+   * @returns {string} imsUserToken - The IMS User access token.
+   * @throws {ErrorWithStatusCode} - If the Authorization header is missing.
+   */
+  function getImsUserToken(headers) {
+    const authorizationHeader = headers.authorization;
+    const BEARER_PREFIX = 'Bearer ';
+    if (!hasText(authorizationHeader)) {
+      throw new ErrorWithStatusCode('Missing Authorization header', STATUS_UNAUTHORIZED);
+    }
+    return authorizationHeader.startsWith(BEARER_PREFIX)
+      ? authorizationHeader.substring(BEARER_PREFIX.length) : authorizationHeader;
   }
 
   /**
    * Create a new API key.
    * @param {Object} requestContext - Context of the request.
-   * @returns {Promise<Response>} 201 Created with the new API key.
+   * @returns {Promise<ApiKey>} - 201 Created with the new API key.
    */
   async function createApiKey(requestContext) {
     const { data, pathInfo: { headers } } = requestContext;
-    const imsOrgId = headers['x-ims-gw-org-id'];
+    const imsOrgId = headers['x-gw-ims-org-id'];
+
+    try {
+      const imsUserToken = getImsUserToken(headers);
+
+      validateRequestData(data);
+      const imsUserProfile = await validateImsOrgId(imsOrgId, imsUserToken);
+
+      // Check if the domains are within the limit. Currently, we only allow one domain per API key.
+      if (data.domains.length > maxDomainsPerApiKey) {
+        throw new ErrorWithStatusCode(`Invalid request: Exceeds the limit of ${maxDomainsPerApiKey} allowed domain(s)`, STATUS_FORBIDDEN);
+      }
+
+      const { authInfo: { profile } } = attributes;
+
+      // Currently the email is assigned as the imsUserId
```

Comment:
I'd recommend a change to the wording:
```suggestion
      // While the property is named 'profile.email', it is in fact the user's IMS User Id
```